src/main/java/io/vertx/core/http/impl/ConnectionManager.java

```diff
@@ -16,139 +16,535 @@
+                }
+                VertxHttp2ClientUpgradeCodec upgradeCodec = new VertxHttp2ClientUpgradeCodec(client.getOptions().getInitialSettings());
+                HttpClientUpgradeHandler upgradeHandler = new HttpClientUpgradeHandler(httpCodec, upgradeCodec, 65536);
+                ch.pipeline().addLast(httpCodec, upgradeHandler, new UpgradeRequestHandler());
+              } else {
+                applyH2ConnectionOptions(pipeline, context);
+              }
+            } else {
+              applyHttp1xConnectionOptions(pipeline, context);
+            }
+          }
+        }
+      });
+      applyConnectionOptions(options, bootstrap);
+      ChannelFuture future = bootstrap.connect(new InetSocketAddress(host, port));
+      future.addListener((ChannelFuture channelFuture) -> {
+        Channel ch = channelFuture.channel();
+        if (channelFuture.isSuccess()) {
+          if (!options.isUseAlpn()) {
+            if (options.isSsl()) {
+              // TCP connected, so now we must do the SSL handshake
+              SslHandler sslHandler = ch.pipeline().get(SslHandler.class);
+              io.netty.util.concurrent.Future<Channel> fut = sslHandler.handshakeFuture();
+              fut.addListener(fut2 -> {
+                if (fut2.isSuccess()) {
+                  http1xConnected(mgr.version, context, port, host, ch, waiter);
+                } else {
+                  handshakeFailure(context, ch, fut2.cause(), waiter);
+                }
+              });
+            } else {
+              if (ch.pipeline().get(HttpClientUpgradeHandler.class) != null) {
+                // Upgrade handler do nothing
+              } else {
+                if (mgr.version == HttpVersion.HTTP_2 && !options.isH2cUpgrade()) {
+                  http2Connected(context, ch, waiter, false);
+                } else {
+                  http1xConnected(mgr.version, context, port, host, ch, waiter);
+                }
+              }
+            }
+          }
+        } else {
+          connectionFailed(context, ch, waiter::handleFailure, channelFuture.cause());
+        }
+      });
+    }
+
+    private void handshakeFailure(ContextImpl context, Channel ch, Throwable cause, Waiter waiter) {
+      SSLHandshakeException sslException = new SSLHandshakeException("Failed to create SSL connection");
+      if (cause != null) {
+        sslException.initCause(cause);
+      }
+      connectionFailed(context, ch, waiter::handleFailure, sslException);
+    }
+
+    private void fallbackToHttp1x(Channel ch, ContextImpl context, HttpVersion fallbackVersion, int port, String host, Waiter waiter) {
+      // Fallback
+      // change the pool to Http1xPool
+      ConnQueue<ClientConnection> http1Queue = qm1x.getConnQueue(address);
+      applyHttp1xConnectionOptions(ch.pipeline(), context);
+      http1Queue.http1xConnected(fallbackVersion, context, port, host, ch, waiter);
+      // Should remove this queue as it may be empty ????
```

Comment:
soon yes :-)

> On Apr 6, 2016, at 10:44 AM, Clement Escoffier <notifications@github.com> wrote:
> 
> In src/main/java/io/vertx/core/http/impl/ConnectionManager.java <https://github.com/eclipse/vert.x/pull/1353#discussion_r58670759>:
> 
> > +
> > +    private void handshakeFailure(ContextImpl context, Channel ch, Throwable cause, Waiter waiter) {
> > +      SSLHandshakeException sslException = new SSLHandshakeException("Failed to create SSL connection");
> > +      if (cause != null) {
> > +        sslException.initCause(cause);
> > +      }
> > +      connectionFailed(context, ch, waiter::handleFailure, sslException);
> > +    }
> > +
> > +    private void fallbackToHttp1x(Channel ch, ContextImpl context, HttpVersion fallbackVersion, int port, String host, Waiter waiter) {
> > +      // Fallback
> > +      // change the pool to Http1xPool
> > +      ConnQueue<ClientConnection> http1Queue = qm1x.getConnQueue(address);
> > +      applyHttp1xConnectionOptions(ch.pipeline(), context);
> > +      http1Queue.http1xConnected(fallbackVersion, context, port, host, ch, waiter);
> > +      // Should remove this queue as it may be empty ????
> any answer to this existential question ?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/eclipse/vert.x/pull/1353/files/bf875485313cf1f6ae199e2ab525539ed5676a2e#r58670759>